tools/eslint-rules/unescaped-regexp-dot.js

```diff
@@ -0,0 +1,119 @@
+'use strict';
+
+const path = require('path');
+const utilsPath = path.join(__dirname, '..', 'eslint', 'lib', 'ast-utils.js');
+const astUtils = require(utilsPath);
+const getLocationFromRangeIndex = astUtils.getLocationFromRangeIndex;
+const getRangeIndexFromLocation = astUtils.getRangeIndexFromLocation;
+
+//------------------------------------------------------------------------------
+// Rule Definition
+//------------------------------------------------------------------------------
+
+module.exports = function(context) {
+  const sourceCode = context.getSourceCode();
+  const regexpStack = [];
+  var regexpBuffer = [];
+  var inRegExp = false;
+
+  function report(node, startOffset) {
+    context.report({
+      node,
+      loc: getLocationFromRangeIndex(
+        sourceCode,
+        getRangeIndexFromLocation(sourceCode, node.loc.start) + startOffset
+      ),
+      message: 'Unescaped dot character in regular expression'
+    });
+  }
+
+  function checkRegExp(nodes) {
+    var escaping = false;
+    var inCharClass = false;
+    for (var n = 0; n < nodes.length; ++n) {
+      const pair = nodes[n];
+      const node = pair[0];
+      const str = pair[1];
+      for (var i = 0; i < str.length; ++i) {
+        switch (str[i]) {
+          case '[':
+            if (!escaping)
+              inCharClass = true;
+            else
+              escaping = false;
+            break;
+          case ']':
+            if (!escaping) {
+              if (inCharClass)
+                inCharClass = false;
+            } else {
+              escaping = false;
+            }
+            break;
+          case '\\':
+            escaping = !escaping;
+            break;
+          case '.':
+            if (!escaping) {
+              if (!inCharClass && (i + 1) < str.length) {
+                switch (str[i + 1]) {
+                  case '+':
+                  case '*':
+                  case '?':
+                  case '{':
+                    break;
+                  default:
+                    report(node, i);
+                }
+              }
+            } else {
+              escaping = false;
+            }
+            break;
+          default:
+            if (escaping)
+              escaping = false;
+        }
+      }
+    }
+  }
+
+  function checkRegExpStart(node) {
+    if (node.callee && node.callee.name === 'RegExp') {
+      if (inRegExp) {
+        regexpStack.push(regexpBuffer);
+        regexpBuffer = [];
+      }
+      inRegExp = true;
+    }
+  }
+
+  function checkRegExpEnd(node) {
+    if (node.callee && node.callee.name === 'RegExp') {
+      checkRegExp(regexpBuffer);
+      if (regexpStack.length) {
+        regexpBuffer = regexpStack.pop();
+      } else {
+        inRegExp = false;
+        regexpBuffer = [];
+      }
+    }
+  }
+
+  return {
+    Literal: function(node) {
+      if (inRegExp && typeof node.value === 'string' && node.value.length)
+        regexpBuffer.push([node, node.value]);
```

Comment:
This could cause some weird behavior if the argument to `RegExp` isn't just string concatenation:

```js
// 'foo.bar' gets interpreted as a regex pattern
new RegExp(function() {
  return crypto.createHash('sha1').update('foo.bar').digest('hex');
}());
```

Or it could handle character classes incorrectly:

```js
const closingBracket = ']';

new RegExp('[abc' + closingBracket + '.');
```

There's no perfect solution to this, so it's probably fine as-is if we're not running into cases like this.